### DISCUSSION TITLE
osrm with shared memory - doubt

### DISCUSSION THREAD

**Discussion by rayudu3745** [Project-OSRM/osrm-backend#6799] [Q&A]
HI, 
I'm running osrm for us-latest with shared data approach as suggested by docs 
and everything works and started without any warning or issue 
but I'm doubtful whether the shared memory is using RAM or DISK 
as df -h shows only 650k used shared memory 
but free -g shows as 37gb

am I missing something here ? pls help 
<img width="623" alt="Screenshot 2024-02-05 at 9 02 30 PM" src="https://github.com/Project-OSRM/osrm-backend/assets/24601586/14478bf9-3557-4a47-9ff1-8d4d3c184b30">
<img width="1254" alt="Screenshot 2024-02-05 at 9 04 36 PM" src="https://github.com/Project-OSRM/osrm-backend/assets/24601586/8a3f8c2b-a118-47b2-9670-84d1eb8db50d">
<img width="1384" alt="Screenshot 2024-02-05 at 9 05 56 PM" src="https://github.com/Project-OSRM/osrm-backend/assets/24601586/74075114-a4dc-4ceb-a2fe-8274efb4916a">
<img width="1269" alt="Screenshot 2024-02-05 at 9 07 19 PM" src="https://github.com/Project-OSRM/osrm-backend/assets/24601586/3b9563d6-e73b-4e3c-8a22-900e39a076b9">

**Comment by rayudu3745** [comment]
can someone help here please

**Comment by mithiridi** [comment]
@rayudu3745 did you get to know anything about this? I have a similar doubt

**Comment by shinomontaz** [comment]
Hi!
At first glance it seems there is no problem in your log.

in case of `shared-mamory` for osrm the underlying mechanism uses Linux's [POSIX shared memory (shm)](https://man7.org/linux/man-pages/man7/shm_overview.7.html), which is backed by RAM only.

`df -h` 650k This is not how much RAM is actually used, but how much virtual space is available for shared memory objects.
The small number can be explained by:
- Lazy allocation:  the kernel reserves space but doesn't "use" it until it's actualy written.
- Kernel accounting: sometimes not all memory usage is reported as 'used' if the objects are delted/unlinked but still mapped by processes.

`free -g` shows the actual RAM ussage  (including that taken up by shared memory segments currently in use).
Is it changed after osrm started (you provides only after start)?

The increase in used RAM after loading with osrm-datastore is a good indicator of in-RAM usage.

Can you run a 
```pmap <PID-of-osrm-routed> | grep shm``` for check mem-mapped files
or check a 
```ipcs -m``` to see shared memory allocations?

**Reply by shinomontaz** [reply]
I've checked locally:

```free -g``` definetely shows RAM usage after osrm-routed started

![osrm-shared-memory-ok](https://github.com/user-attachments/assets/a1f89882-e9c5-4516-b984-2ec35508efb2)

pmap also shows a shared memory usage:

```
pmap 302227 | grep shm
00007fb5b7441000 3084720K r--s-   [ shmid=0x5 ]
00007fb6738ad000 1123048K r--s-   [ shmid=0x4 ]
```

So, it seems all is ok
